notify/silences.go

```diff
@@ -124,3 +129,54 @@ func (am *GrafanaAlertmanager) DeleteSilence(silenceID string) error {
 
 	return nil
 }
+
+func (am *GrafanaAlertmanager) SilenceState() (SilenceState, error) {
+	r, w := io.Pipe()
+	go func() {
+		_, err := am.silences.Snapshot(w)
+		_ = w.CloseWithError(err)
+	}()
+
+	// Trade-off between type safety and performance ahead.
+	// This is a bit awkward as we marshalled the silence just to unmarshalling it again. We could keep the return value
+	// as (string, error) and return the string directly as, for now, callers just needs the string itself.
+	/// However, this would remove type safety on the interface, forcing the caller to trust the AM implementation to
+	// always return a consistent type or to perform the unmarshalling themselves.
+	return DecodeState(r)
+}
+
+// SilenceState copied from state in prometheus-alertmanager/silence/silence.go.
+type SilenceState map[string]*silencepb.MeshSilence
+
+func (s SilenceState) MarshalBinary() ([]byte, error) {
+	var buf bytes.Buffer
+
+	for _, e := range s {
+		if _, err := pbutil.WriteDelimited(&buf, e); err != nil {
+			return nil, err
+		}
+	}
+	return buf.Bytes(), nil
+}
+
+// DecodeState copied from decodeState in prometheus-alertmanager/silence/silence.go.
+func DecodeState(r io.Reader) (SilenceState, error) {
+	st := SilenceState{}
+	for {
+		var s silencepb.MeshSilence
+		_, err := pbutil.ReadDelimited(r, &s)
+		if err == nil {
+			if s.Silence == nil {
+				return nil, silence.ErrInvalidState
+			}
+			st[s.Silence.Id] = &s
+			continue
+		}
+		//nolint:errorlint
+		if err == io.EOF {
```

Comment:
This should be `errors.Is`? https://github.com/prometheus/alertmanager/blob/main/silence/silence.go#L964

Reply:
Huh, nice catch! Guess we updated the AM sometime between when I copied that code and now.